### PR TITLE
Adding sub dialogue for formatting of summary tables

### LINKED ISSUE #6770: Formatting table output for mmtable2
We have introduced `mmtable2` into R-Instat to create summary tables. The package works similar to `ggplot2` where layers are added through `+` to enhance or change the table.

There is a vignette [here](https://ianmoran11.github.io/mmtable2/articles/Some-minimal-examples.html) which has three sections:
* Adding headers  - we have done this already in the summary and frequency tables dialogs.
* Combining tables  - looks exciting and I'm sure will be introduced into R-Instat in the future!
* Formatting tables  - what we are interested in for this issue

Formatting tables allows changes to be made to the mmtable2 object using the `gt` package. There is some sample code for formatting tables at the bottom of this comment, but it looks like there are many different areas where we can format the tables.

## Formatting Tables

`cells_format` - this is to format the cells in the tables
`header_format` - this is to format the headers in the tables
`table_format` - this is to format the `mmtable2` object altogether

For these functions 

* The first parameter states what to format
E.g. for `header_format`, which header to format.
E.g.2. for `table_format`, `cells_body` from the `gt` package can be read in to specify which rows or columns or cells to format.

* The second parameter they can take is `style` which specifies how to format it.
E.g. `style = list(cell_text(...))` where `cell_text` is from the `gt` package.

From a quick look, `cell_borders` and `cell_fill` are other parameters that can be read into `style = list(...)`. 

So one task is to explore what, if any, other parameters can be read into the `style = list(...)` function from the `gt` package. A quick glance at the `tab_style` help page in the `gt` package suggests it is only these three, however perhaps there are more that I have missed.

Another task is to then look at what parameters these functions can take.
That is, what parameters `cell_text` can take in R, and similarly so for `cell_fill` and `cell_borders`.

## Titles, Captions, Etc.
The `table_source_note` function is another layer that can be added. This takes arguments from `gt::tab_source_note`, so the parameter this takes is `source_note` with the value as a string of what the caption should say.

From a quick look of the `gt` package, `tab_header` is another layer that can add a title and subtitle to the `mmtable2` object. 

A third task is what else from the `gt` package works as a new layer for an `mmtable2` package. It might be worth looking at the [vignette](https://cran.r-project.org/web/packages/gt/gt.pdf).
From a quick google, https://gt.rstudio.com/articles/intro-creating-gt-tables.html discusses other bits to add (such as footnotes) so it could be worth having a look there. 

## Other Options 
`header_merged_cols()` is another option in the `mmtable2` function to edit the headers
Are there any other options?

Finally, we then want to take all of these bits and plan a dialog or sub-dialog to format the `mmtable2` object in R-Instat.

## R Code
```
#remotes::install_github("ianmoran11/mmtable2")

library(gapminder)
library(dplyr)
library(tidyr)
library(stringr)
library(purrr)
library(gt)
library(mmtable2)

gapminder_mm %>% 
  filter(var != "Life expectancy") %>% 
  mmtable(cells = value) +
  header_top(year) +
  header_left(country) +
  header_top_left(var)  +
  header_left_top(continent)  +
  cells_format(cell_predicate = T, style = list(cell_text(align = "right"))) +
  header_format(header = year, style = list(cell_text(align = "right"))) +
  header_format("all_cols", style = list(cell_text(weight = "bolder"))) +
  header_format("all_rows", style = list(cell_text(weight = "bolder"))) +
  table_format(
    locations = cells_body(rows = c(1,3,5,7,9,11)),
    style = list(cell_borders(sides = "top",color = "grey"))) + 
  table_source_note(source_note = "Excerpt of the Gapminder data on life expectancy, GDP per capita, and population by country." )
```

(As a side note, I can see some functions called `cols_hide` and bits in the `gt` package, perhaps that would be an interesting avenue to explore later for editing mmtable2 objects. However, that is not to do with formatting).

### PR DESCRIPTION
**Pull request by Ivanluv**
@africanmathsinitiative/developers this fixes  #6770 and replaces PR #6977

### CODE REVIEW

**Comment by lloyddewit** [comment]
@Ivanluv Thanks, is this PR ready for test/review or are you still working on it?
(if not ready, then you could set this PR to 'draft' to make its status clearer)

**Comment by Ivanluv** [comment]
@lloyddewit thank you for the reminder.@africanmathsinitiative/developers  this is ready for review

**Review by rdstern — CHANGES_REQUESTED** [review]
I keep having more questions, but if @lilyclements agrees, I would like to merge this as soon as possible.  So just one thing to check - hope it is easy.  
This is that the order of the variables in the table is not the same as the order in the receiver.  I think it is alphabetical in the table.  I can live with this temporarily if it is difficult to fix?
My next question (for a new issue?) is how to dictate the order of the summaries in relation to the variables and the other factors in the table?

**Comment by lilyclements** [comment]
> This is that the order of the variables in the table is not the same as the order in the receiver

This is fixed in PR #7128

**Comment by lilyclements** [comment]
> I keep having more questions, but if @lilyclements agrees, I would like to merge this as soon as possible

@rdstern I agree that this is good to merge, and you can outlined your questions elsewhere to work on

**Review by rdstern — DISMISSED** [review]
@lloyddewit or @shadrackkibet there is still a lot to do on improving the tables.  But this is already a major improvement on 7.3.1 and I suggest it be merged, and then we make another issue for further improvements.  I have checked and @lilyclements agrees to merga at this point.

**Comment by lloyddewit** [comment]
@HawardKetoyoMsatsi Please could you do the first review? thanks

File: instat/dlgFrequency.vb [hunk 1]
```diff
@@ -21,8 +22,14 @@ Public Class dlgFrequency
     Private clsDefaultFunction As New RFunction
     Private clsMmtableFunction As New RFunction
     Private clsFrequencyOperator As New ROperator
-    Private clsMmtableOperator As New ROperator
```
Comment (HawardKetoyoMsatsi) [inline instat/dlgFrequency.vb:24]: Remove this line, clsMmtableOperator has been initialized twice

**Comment by HawardKetoyoMsatsi** [comment]
image

The design of the sub dialog looks good, with just a few enhancements.
Add (:) after `Title Font` and `Title Size` labels.
The Minimum value of the input `Title Size:` is set to zero, Could it have been a positive value?

**Review by HawardKetoyoMsatsi — CHANGES_REQUESTED** [review]

**Comment by Ivanluv** [comment]
> The Minimum value of the input `Title Size:` is set to zero, Could it have been a positive value?

I have that set to 1 now

File: instat/dlgFrequency.vb (hunk 1 above)
Comment (Ivanluv) [reply instat/dlgFrequency.vb:24]: I cant seem to find the other initialization

Comment (HawardKetoyoMsatsi) [reply instat/dlgFrequency.vb:24]: Check clsMmtableOperator appears in both lines 24 and 25.

**Comment by Ivanluv** [comment]
@HawardKetoyoMsatsi are you done reviewing this?

**Comment by HawardKetoyoMsatsi** [comment]
Thanks, @Ivanluv I see the changes I requested have been made. 
@lloyddewit over to you.

File: instat/dlgSummaryTables.vb [hunk 2]
```diff
@@ -298,32 +426,31 @@ Public Class dlgSummaryTables
...
                     clsHeaderLeftFunction.AddParameter("variable", strcolumn, iPosition:=0)
                     clsColumnOperator.AddParameter(strcolumn, clsRFunctionParameter:=clsHeaderLeftFunction, iPosition:=iColumn)
                 Else
+                    Dim clsHeaderTopFunction As New RFunction
+                    clsHeaderTopFunction.SetPackageName("mmtable2")
+                    clsHeaderTopFunction.SetRCommand("header_left_top")
                     clsHeaderTopFunction.AddParameter("variable", strcolumn, iPosition:=0)
                     clsColumnOperator.AddParameter(strcolumn, clsRFunctionParameter:=clsHeaderTopFunction, iPosition:=iColumn)
                 End If
                 iColumn = iColumn + 1
```
Comment (lloyddewit) [inline instat/dlgSummaryTables.vb:451]: shorten and make more readable?
```suggestion
                Dim clsHeaderFunction As New RFunction
                clsHeaderFunction.SetPackageName("mmtable2")
                clsHeaderFunction.SetRCommand(If(iColumn < iNumberOfColumns, "header_top_left", "header_left_top"))
                clsHeaderFunction.AddParameter("variable", strColumn, iPosition:=0)
                clsColumnOperator.AddParameter(strColumn, clsRFunctionParameter:=clsHeaderFunction, iPosition:=iColumn)
                iColumn += 1
```

File: instat/dlgSummaryTables.vb [hunk 3]
```diff
@@ -298,32 +426,31 @@ Public Class dlgSummaryTables
...
-
-        clsHeaderTopFunction.SetPackageName("mmtable2")
-        clsHeaderTopFunction.SetRCommand("header_left_top")
-
         clsColumnOperator.ClearParameters()
+        clsMutableOperator.RemoveParameterByName("columnOp")
 
         If Not ucrReceiverFactors.IsEmpty AndAlso ucrNudColumnFactors.GetText() <> "" Then
             iNumberOfColumns = ucrNudColumnFactors.GetText()
             For Each strcolumn As String In ucrReceiverFactors.GetVariableNamesAsList
```
Comment (lloyddewit) [inline instat/dlgSummaryTables.vb:437]:
```suggestion
            For Each strColumn As String In ucrReceiverFactors.GetVariableNamesAsList
```

File: instat/sdgFormatSummaryTables.vb [hunk 4]
```diff
@@ -1,3 +1,710 @@
...
+' This program is distributed in the hope that it will be useful,
+' but WITHOUT ANY WARRANTY; without even the implied warranty of
+' MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+' GNU General Public License for more details.
+'
+' You should have received a copy of the GNU General Public License 
+' along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
+Imports instat.Translations
+Public Class sdgFormatSummaryTables
```
Comment (lloyddewit) [inline instat/sdgFormatSummaryTables.vb:18]:
```suggestion

Public Class sdgFormatSummaryTables
```

File: instat/sdgFormatSummaryTables.vb [hunk 5]
```diff
@@ -1,3 +1,710 @@
...
+    Private clsTableTitleFunction, clsTabFootnoteTitleFunction, clsTableSourcenoteFunction, clsFootnoteCellFunction,
+        clsCellTextFunction, clsCellBorderFunction, clsCellFillFunction, clsHeaderFormatFunction, clsFootnoteTitleLocationFunction,
+        clsTabOptionsFunction, clsBorderWeightPxFunction, clsFootnoteSubtitleLocationFunction, clsTabFootnoteSubtitleFunction, clsFootnoteCellBodyFunction,
+        clsStyleListFunction, clsStubHeadFunction, clsSecondFootnoteCellFunction, clsSecondFootnoteCellBodyFunction,
+        clsTabStyleFunction, clsTabStyleCellTextFunction, clsTabStylePxFunction As New RFunction
+    'The dummy Function is used by input controls that add parameter manually,
+    'when openning the subdialogue from multiple dialogues
```
Comment (lloyddewit) [inline instat/sdgFormatSummaryTables.vb:25]:
```suggestion
    'The dummy Function is used by input controls that add the parameter manually,
    'when opening the subdialogue from multiple dialogues
```

File: instat/sdgFormatSummaryTables.vb [hunk 6]
```diff
@@ -1,3 +1,710 @@
...
+    Private clsDummyFunction As New RFunction
+    Private clsPipeOperator, clsMutableOperator, clsJoiningOperator, clsTabFootnoteOperator As New ROperator
+    Private bControlsInitialised As Boolean = False
+    Private bRCodeSet As Boolean = False
+    Private Sub sdgFormatSummaryTables_load(sender As Object, e As EventArgs) Handles MyBase.Load
+        autoTranslate(Me)
+    End Sub
+    Public Sub InitialiseControls()
```
Comment (lloyddewit) [inline instat/sdgFormatSummaryTables.vb:33]:
```suggestion

    Public Sub InitialiseControls()
```

File: instat/sdgFormatSummaryTables.vb [hunk 7]
```diff
@@ -1,3 +1,710 @@
...
+        ucrInputSubtitle.SetLinkedDisplayControl(lblSubtitle)
+
+        ucrInputTitleFootnote.SetParameter(New RParameter("title_footnote", iNewPosition:=2))
+
+        ucrChkTitleFootnote.SetText("Add title footnote")
+        ucrChkTitleFootnote.AddToLinkedControls(ucrInputTitleFootnote, {True}, bNewLinkedHideIfParameterMissing:=True)
+        ucrChkTitleFootnote.SetLinkedDisplayControl(grpTitleSubtitle)
+
+        ucrChkTitleFootnote.AddParameterPresentCondition(True, "titlefootnote")
+        ucrChkTitleFootnote.AddParameterPresentCondition(False, "titlefootnote", False)
```
Comment (lloyddewit) [inline instat/sdgFormatSummaryTables.vb:52]: `"titlefootnote"` is different to the parameter set on line 45 (`"title_footnote"`). 
The same is true on other calls to `AddParameterPresentCondition` below.
Does this matter?
Sorry if this is an ignorant question

File: instat/sdgFormatSummaryTables.vb [hunk 8]
```diff
@@ -1,3 +1,710 @@
...
+        ucrInputCellFootnote.SetParameter(New RParameter("cellFootnote", iNewPosition:=4))
+        ucrInputCellFootnote.SetLinkedDisplayControl(lblFootnote)
+
+        ucrChkAddFootnote.SetText("Add footnote")
+        ucrChkAddFootnote.AddToLinkedControls({ucrInputCellFootnote, ucrInputFootnoteColumnLocation, ucrInputFootnoteRowLocation}, {True}, bNewLinkedHideIfParameterMissing:=True)
+
+        ucrChkAddFootnote.AddParameterPresentCondition(True, "cellfootnote")
+        ucrChkAddFootnote.AddParameterPresentCondition(False, "cellfootnote", False)
+
+        ucrChKAddSecondFootnote.SetText("Add Second footnote")
```
Comment (lloyddewit) [inline instat/sdgFormatSummaryTables.vb:93]:
```suggestion
        ucrChKAddSecondFootnote.SetText("Add second footnote")
```

File: instat/sdgFormatSummaryTables.vb [hunk 9]
```diff
@@ -1,3 +1,710 @@
...
+        ucrInputSecondCellFootnote.SetLinkedDisplayControl(lblSecondFootnote)
+
+        ucrInputSecondFootnoteColumnLocation.SetLinkedDisplayControl(grpSecondFootnote)
+
+        ucrChKAddSecondFootnote.AddParameterPresentCondition(True, "second_cellfootnote")
+        ucrChKAddSecondFootnote.AddParameterPresentCondition(False, "second_cellfootnote", False)
+
+        ucrInputAddSourceNote.SetParameter(New RParameter("source_note", iNewPosition:=8))
+
+        ucrChkAddSourcenote.SetText("Add Source")
```
Comment (lloyddewit) [inline instat/sdgFormatSummaryTables.vb:112]:
```suggestion
        ucrChkAddSourcenote.SetText("Add source")
```

File: instat/sdgFormatSummaryTables.vb [hunk 10]
```diff
@@ -1,3 +1,710 @@
...
+        ucrChkAddSourcenote.AddToLinkedControls(ucrInputAddSourceNote, {True}, bNewLinkedHideIfParameterMissing:=True)
+
+        ucrChkAddSourcenote.AddParameterPresentCondition(True, "source_note")
+        ucrChkAddSourcenote.AddParameterPresentCondition(False, "source_note", False)
+        'Header
+        ucrChkAddHeader.SetText("Add Header")
```
Comment (lloyddewit) [inline instat/sdgFormatSummaryTables.vb:118]:
```suggestion
        ucrChkAddHeader.SetText("Add header")
```

File: instat/sdgFormatSummaryTables.vb [hunk 11]
```diff
@@ -1,3 +1,710 @@
...
+        ucrChkAddHeader.AddToLinkedControls({ucrChkStyleText, ucrPnlHeader}, {True}, bNewLinkedHideIfParameterMissing:=True)
+        ucrPnlHeader.SetLinkedDisplayControl(lblHeader)
+        ucrChkAddHeader.AddParameterPresentCondition(True, "header")
+        ucrChkAddHeader.AddParameterPresentCondition(False, "header", False)
+
+        ucrChkAddTableFormat.SetText("Add Table format")
```
Comment (lloyddewit) [inline instat/sdgFormatSummaryTables.vb:124]:
```suggestion
        ucrChkAddTableFormat.SetText("Add table format")
```

File: instat/sdgFormatSummaryTables.vb [hunk 12]
```diff
@@ -1,3 +1,710 @@
...
+        ucrChkAddTableFormat.AddParameterPresentCondition(True, "table_format")
+        ucrChkAddTableFormat.AddParameterPresentCondition(False, "table_format", False)
+
+        ucrPnlHeader.SetLinkedDisplayControl(lblHeader)
+        ucrPnlHeader.SetParameter(New RParameter("header", iNewPosition:=0))
+        ucrPnlHeader.AddRadioButton(rdoAllColumns, Chr(34) & "all_cols" & Chr(34))
+        ucrPnlHeader.AddRadioButton(rdoAllRows, Chr(34) & "all_rows" & Chr(34))
+        ucrPnlHeader.SetLinkedDisplayControl(lblHeader)
+
+        ucrChkStyleText.SetText("Add Text style")
```
Comment (lloyddewit) [inline instat/sdgFormatSummaryTables.vb:136]:
```suggestion
        ucrChkStyleText.SetText("Add text style")
```

File: instat/sdgFormatSummaryTables.vb [hunk 13]
```diff
@@ -1,3 +1,710 @@
...
+        ucrChkAddTitleSubtitle.AddParameterPresentCondition(False, "title\subtitle", False)
+
+        ucrInputFootnoteColumnLocation.SetParameter(New RParameter("col_location", 5))
+        ucrInputFootnoteColumnLocation.SetValidationTypeAsNumericList()
+
+        ucrInputFootnoteRowLocation.SetParameter(New RParameter("row_location", iNewPosition:=6))
+        ucrInputFootnoteRowLocation.SetValidationTypeAsNumericList()
+        ucrInputFootnoteRowLocation.SetLinkedDisplayControl(grpLocation)
+
+        ucrInputCellFootnote.SetParameter(New RParameter("cellFootnote", iNewPosition:=4))
```
Comment (lloyddewit) [inline instat/sdgFormatSummaryTables.vb:84]: This parameter name has a different format than the others (I expected "cell_footnote"), is it correct?

File: instat/sdgFormatSummaryTables.vb [hunk 14]
```diff
@@ -1,3 +1,710 @@
...
+
+        ucrNudStyleTextIndent.SetLinkedDisplayControl(lblStyleTextIndent)
+        ucrNudStyleTextIndent.Increment = 1
+
+        ucrChkStyleBoarder.SetText("Add style border")
+        ucrChkStyleBoarder.AddToLinkedControls({ucrInputStyleBorderSides, ucrInputStyleBorderColor, ucrInputStyleBorderStyle, ucrNudStyleBorderWeight}, {True},
+                                                bNewLinkedHideIfParameterMissing:=True)
+
+        ucrChkStyleBoarder.AddParameterPresentCondition(True, "boarderStyle")
+        ucrChkStyleBoarder.AddParameterPresentCondition(False, "boarderStyle", False)
```
Comment (lloyddewit) [inline instat/sdgFormatSummaryTables.vb:262]: should it be 'border' not 'boarder'?

**Review by lloyddewit — CHANGES_REQUESTED** [review]

File: instat/sdgFormatSummaryTables.vb (hunk 13 above)
Comment (Ivanluv) [reply instat/sdgFormatSummaryTables.vb]: Yes, it should be `cell_footnote`

File: instat/sdgFormatSummaryTables.vb (hunk 14 above)
Comment (Ivanluv) [reply instat/sdgFormatSummaryTables.vb]: `border`...I have corrected that

**Comment by Ivanluv** [comment]
@lloyddewit could you have a look again

**Comment by lloyddewit** [comment]
@Ivanluv thank you for all the changes, I'm happy with all the comments you marked as resolved.
There are 3 unresolved comments above (I had to click on a link in the list to unhide them, so maybe you didn't see them), please could you try and resolve these also?
Thanks

File: instat/sdgFormatSummaryTables.vb (hunk 7 above)
Comment (Ivanluv) [reply instat/sdgFormatSummaryTables.vb]: I have made the changes

File: instat/sdgFormatSummaryTables.vb (hunk 8 above)
Comment (Ivanluv) [reply instat/sdgFormatSummaryTables.vb:95]: Done

File: instat/sdgFormatSummaryTables.vb (hunk 9 above)
Comment (Ivanluv) [reply instat/sdgFormatSummaryTables.vb:114]: Done

**Comment by Ivanluv** [comment]
@lloyddewit I missed the three comments but they are now resolved. Could you have a look?

**Comment by lloyddewit** [comment]
> @lloyddewit I missed the three comments but they are now resolved. Could you have a look?

@Ivanluv thanks for fixing those, I set them to resolved

**Review by lloyddewit — APPROVED** [review]

**Comment by lloyddewit** [comment]
@rdstern Please could you test again?
Some of the parameters have changed since your last approval, so it may be worth a full test.
Thanks!

**Comment by rdstern** [comment]
@Ivanluv just checking.  I get the right answers here, but the order is still alphabetical of the variables, and not the order they were in the dialogue.  I am pretty sure this was fixed by @lilyclements so assume it might be ok once merged?  Am I right?
image

And then it gets more "interesting" when I add the margins!  This might be a @lilyclements question?
image

That's from the dialogue with the R-Instat review like this:
image

The R-code is here:

image

And I haven't got to the sub-dialogues yet!

**Comment by rdstern** [comment]
On the sub-dialogues I was able to play and got this far:
image

I had hoped to put some colour into the table itself, but didn't manage.  I also wrote some text as a "stub" but couldn't find it.  And I tried to align the numbers above as left and they still seem right!

The code for the above is here:
image

Overall we always knew the formatting is "work-in-progress" and I'd be happy for these to be merged as is, once the main part is examined.

It would be really good to get this into the forthcoming release, because the existing version has a serious bug.

**Comment by Ivanluv** [comment]
> @Ivanluv just checking. I get the right answers here, but the order is still alphabetical of the variables, and not the order they were in the dialogue. I am pretty sure this was fixed by @lilyclements so assume it might be ok once merged? Am I right?

True this is fixed, I had not updated the branch but now it should be okay

**Review by lloyddewit — APPROVED** [review]

**Comment by lloyddewit** [comment]
@rdstern If you can approve, then we can merge, thanks

**Comment by lilyclements** [comment]
> And then it gets more "interesting" when I add the margins! This might be a @lilyclements question? image

@rdstern thanks - I think this should be now fixed in PR #7172 (it is working from my side now anyway).

**Comment by lilyclements** [comment]
> And then it gets more "interesting" when I add the margins! This might be a @lilyclements question? image

@rdstern thanks - I think this should be now fixed in PR #7172 (it is working from my side now anyway).

**Comment by lloyddewit** [comment]
@Ivanluv @lilyclements As you know, we're hoping to build a new release on Monday. 
I understand from @rdstern that: 

'PR #7107 and PR #7172 are (together) the important ones.  Together they are almost ok now, but I am not sure if they will both merge, or Ivan could incorporate 7172 in his 7107.  There also remains a very minor issue I found, that if variable names contain a ".", then it gives an error.  Lily is checking, but we could merge anyway if she doesn't find it.'

If there's any chance of making progress today on these PRs, then that would be great.
I'm available this morning and then travelling this afternoon/evening. I'll check again on Sunday to see if there's anything I can review/approve/merge.

Thanks

**Comment by Ivanluv** [comment]
@lloyddewit   I could incorparate PR #7172 into PR #7107 then have #7172 closed to avoid conflicts

**Comment by lloyddewit** [comment]
> @lloyddewit I could incorparate PR #7172 into PR #7107 then have #7172 closed to avoid conflicts

@Ivanluv Sounds sensible. If you think this is the best approach, then please go ahead.
Will you or @lilyclements still try and resolve the '.' issue first, or do you prefer to merge and solve the issue in a separate PR?

**Comment by lilyclements** [comment]
> > @lloyddewit I could incorparate PR #7172 into PR #7107 then have #7172 closed to avoid conflicts
> 
> @Ivanluv Sounds sensible. If you think this is the best approach, then please go ahead. Will you or @lilyclements still try and resolve the '.' issue first, or do you prefer to merge this PR first and solve the issue in a new PR?

I will look into this issue today and it is easier to do it on a new PR.

**Pull request merged**